rust/lance-index/src/scalar/inverted/builder.rs

```diff
@@ -0,0 +1,361 @@
+    docs: &mut DocSet,
+) -> Result<()> {
+    let mut tokenizer = TOKENIZER.clone();
+    let mut stream = new_data;
+    while let Some(batch) = stream.try_next().await? {
+        let doc_col = batch.column(0).as_string::<Offset>();
+        let row_id_col = batch[ROW_ID].as_primitive::<datatypes::UInt64Type>();
+
+        for (doc, row_id) in doc_col.iter().zip(row_id_col.iter()) {
+            let doc = doc.unwrap();
+            let row_id = row_id.unwrap();
+            let mut token_stream = tokenizer.token_stream(doc);
+            let mut row_token_cnt = HashMap::new();
+            let mut token_cnt = 0;
+            while let Some(token) = token_stream.next() {
+                let token_id = token_set.add(token.text.to_owned());
+                row_token_cnt
+                    .entry(token_id)
+                    .and_modify(|cnt| *cnt += 1)
+                    .or_insert(1);
+                token_cnt += 1;
+            }
+            invert_list.add(row_token_cnt, row_id);
+            docs.add(row_id, token_cnt);
+        }
+    }
+
+    Ok(())
+}
+
+// InvertedList is a mapping from token ids to row ids
+// it's used to retrieve the documents that contain a token
+#[derive(Debug, Clone, Default, DeepSizeOf)]
+pub struct InvertedList {
+    // the index is the token id
+    inverted_list: Vec<PostingList>,
+}
+
+impl InvertedList {
+    // the schema of the inverted list is | row_id | frequency |
+    // and store the offset of
+    pub fn to_batch(&self) -> Result<RecordBatch> {
+        let batches = self
+            .inverted_list
+            .iter()
+            .map(|list| list.to_batch())
+            .collect::<Result<Vec<_>>>()?;
+
+        let offsets = batches
+            .iter()
+            .scan(0, |offset, batch| {
+                *offset += batch.num_rows();
+                Some(*offset - batch.num_rows())
+            })
+            .collect_vec();
+        let metadata = HashMap::from_iter(vec![(
+            "offsets".to_owned(),
+            serde_json::to_string(&offsets)?,
+        )]);
+
+        let batch =
+            concat_batches(batches[0].schema_ref(), batches.iter())?.with_metadata(metadata)?;
+        Ok(batch)
+    }
+
+    pub async fn load(reader: Arc<dyn IndexReader>) -> Result<Self> {
+        let mut inverted_list = Vec::with_capacity(reader.num_rows());
+        for i in 0..reader.num_batches().await {
+            let batch = reader.read_record_batch(i).await?;
+            let row_ids_col = batch[ROW_ID].as_list::<i32>();
+            let frequencies_col = batch[FREQUENCY_COL].as_list::<i32>();
+
+            for (row_ids, frequencies) in row_ids_col.iter().zip(frequencies_col.iter()) {
+                let row_ids = row_ids.unwrap();
+                let frequencies = frequencies.unwrap();
+                let row_ids = row_ids.as_primitive::<UInt64Type>().values().to_vec();
+                let frequencies = frequencies.as_primitive::<Float32Type>().values().to_vec();
```

Comment:
It seems like you are doing `O(n_tokens)` mem copies, which seems very inefficient. Have you considered making the `PostingList` storing the arrays directly? Or a [ScalarBuffer](https://arrow.apache.org/rust/arrow/buffer/struct.ScalarBuffer.html)?

Reply:
Here I don't store them as array because i saw it's much slower than `Vec`, can try scalar buffer and switch to it if there's no obvious perf impact